aws/helper_functions.py

```diff
@@ -116,19 +71,76 @@ def download_best_model( bucket, s3_model_output, hpo_results, local_directory )
     """ Download best model from S3"""
     try:
         target_bucket = boto3.resource('s3').Bucket( bucket )
-        path_prefix = s3_model_output.split('/')[3] + '/' + hpo_results['BestTrainingJob']['TrainingJobName'] + '/output/'
+        path_prefix = s3_model_output.split('/')[-1] + '/' + hpo_results['BestTrainingJob']['TrainingJobName'] + '/output/'
         objects = target_bucket.objects.filter( Prefix = path_prefix )    
         for obj in objects:                
             path, filename = os.path.split( obj.key )
             local_filename = local_directory + '/' + 'best_' + filename
-            full_url = 's3://' + bucket + '/' + path_prefix + obj.key
+            s3_path_to_model = 's3://' + bucket + '/' + path_prefix + filename
             target_bucket.download_file( obj.key, local_filename )
             print(f'Successfully downloaded best model\n'
                   f'> filename: {local_filename}\n'
                   f'> local directory : {local_directory}\n\n'
-                  f'full S3 path : {full_url}')
-        return local_filename
+                  f'full S3 path : {s3_path_to_model}')
+        return local_filename, s3_path_to_model
     
     except Exception as download_error:
         print( f'! Unable to download best model: {download_error}')
-        return None
\ No newline at end of file
+        return None
+    
+def new_job_name_from_config ( dataset_directory, region, code_choice, 
+                               algorithm_choice, cv_folds,
+                               instance_type, trim_limit = 32 ):
+    """ 
+    Build a jobname string that captures the HPO configuration options.
+    This is helpful for intepreting logs and for general book-keeping
+    """
+    job_name = None    
+    try:
+        if dataset_directory in [ '1_year', '3_year', '10_year']:
+            data_choice_str = 'air'
+            validate_region ( region )
+        elif dataset_directory in [ 'NYC_taxi']:
+            data_choice_str = 'nyc'
+            validate_region ( region )
+        else:
+            data_choice_str = 'byo'
+            
+        code_choice_str = code_choice[0] + code_choice[-3:]
+        
+        if 'randomforest' in algorithm_choice.lower() : algorithm_choice_str = 'RF'
+        if 'xgboost' in algorithm_choice.lower() : algorithm_choice_str = 'XGB'    
+        
+        instance_type_str = '-'.join( instance_type.split('.')[1:] )        
+
+        random_8char_str = ''.join( random.choices( uuid.uuid4().hex, k=8 ) )        
```

Comment:
Are we just looking to construct a unique job name? If so, I'd keep the full uuid. If not, is there any reason to think that randomly choosing 8 characters from the uuid is better than merely truncating it to 8 characters?

Reply:
I think this was to stay within the 32 character limit of SageMaker job names.

Reply:
Sorry, let me rephrase with better context: it looks like we're already trimming down to 32 characters at line 122. Is there any reason to "pre-trim" this portion of the name?

Also, same question about the random selection of characters from the generated uuid.